src/EditorFeatures/Core/EditorConfigSettings/Updater/SettingsUpdateHelper.cs

```diff
@@ -348,5 +423,93 @@ static bool IsBothVisualBasicAndCSharp(Language language, string[] splicedFileEx
             var result = editorConfigText.WithChanges((TextChange)new TextChange(new TextSpan(editorConfigText.Length, 0), prefix + newEntry));
             return (result, lastValidHeaderSpanEnd, result.Lines[^2]);
         }
+
+        private static string GetStringForNewEntry(string optionName, string optionValue, ImmutableArray<string> ids, string severity)
+        {
+            var sb = new StringBuilder($"{optionName}={optionValue}\r\n");
+            var orderedIds = ids.OrderBy(StringComparer.OrdinalIgnoreCase).ToArray();
+            foreach (var id in orderedIds.Take(orderedIds.Length - 1))
+            {
+                sb.AppendLine($"dotnet_diagnostic.{id}.severity={severity}");
+            }
+            // do not append an additional newline at the end of mult-line options
+            sb.Append($"dotnet_diagnostic.{orderedIds.Last()}.severity={severity}");
+            return sb.ToString();
+        }
+
+        private static (string untrimmedKey, string key, string value, string severitySuffixInValue, string commentValue) GetOptionParts(string? curLineText)
+        {
+            var groups = s_optionEntryPattern.Match(curLineText).Groups;
+            var untrimmedKey = groups[1].Value.ToString();
+            var key = untrimmedKey.Trim();
+            var value = groups[2].Value.ToString();
+            var severitySuffixInValue = groups[3].Value.ToString();
+            var commentValue = groups[4].Value.ToString();
+            return (untrimmedKey, key, value, severitySuffixInValue, commentValue);
+        }
+
+        private static bool IsValidHeader(TextLine? mostRecentHeader, TextLine? lastValidHeader)
+        {
+            return mostRecentHeader is not null &&
+                   lastValidHeader is not null &&
+                   mostRecentHeader.Equals(lastValidHeader);
+        }
+
+        private static bool ShouldSetAsLastValidHeader(string curLineText, out string mostRecentHeaderText)
+        {
+            var groups = s_headerPattern.Match(curLineText.Trim()).Groups;
+            mostRecentHeaderText = groups[1].Value.ToString().ToLowerInvariant();
+            return mostRecentHeaderText.Equals("*", StringComparison.Ordinal);
+        }
+
+        private static (string fileName, string[] splicedFileExtensions) GetHeaderParts(string mostRecentHeaderText)
+        {
+            // We splice on the last occurrence of '.' to account for filenames containing periods.
+            var nameExtensionSplitIndex = mostRecentHeaderText.LastIndexOf('.');
+            var fileName = mostRecentHeaderText.Substring(0, nameExtensionSplitIndex);
+            var splicedFileExtensions = mostRecentHeaderText[(nameExtensionSplitIndex + 1)..].Split(',', ' ', '{', '}');
+
+            // Replacing characters in the header with the regex equivalent.
+            fileName = fileName.Replace(".", @"\.");
+            fileName = fileName.Replace("*", ".*");
+            fileName = fileName.Replace("/", @"\/");
```

Comment:
this feels unsafe.  lots of other filename characters would violate regex semantics.

Reply:
every editorconfig parsers I am aware of also use regex for this [including the one in the compiler](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.SectionNameMatching.cs).  The correct thing to do is to unify all this parser logic with the compilers implementation.